### PR TITLE
perf: Generate SQL with fewer CTEs

### PR DESCRIPTION
**Pull request by TrevorBergeron**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: bigframes/core/compile/compiled.py [hunk 1]
```diff
@@ -355,6 +346,9 @@ def _to_ibis_expr(
         return table
 
     def filter(self, predicate: ex.Expression) -> UnorderedIR:
+        if any(map(is_window, map(self._get_ibis_column, predicate.unbound_variables))):
```
Comment (tswast) [inline bigframes/core/compile/compiled.py:349]: Is this a fix for something? If so, maybe worth a mention in the changelog and corresponding test?

File: bigframes/core/compile/compiled.py [hunk 2]
```diff
@@ -785,15 +779,32 @@ def promote_offsets(self, col_id: str) -> OrderedIR:
...
+            expr_builder = self.builder()
+            expr_builder.columns = [
+                self._compile_expression(
+                    ordering.total_order_col.scalar_expression
+                ).name(col_id),
+                *self.columns,
+            ]
+            return expr_builder.build()
+        # Cannot nest analytic expressions, so reproject to cte first if needed.
+        # Also ibis cannot window literals, so need to reproject those (even though this is legal in googlesql)
```
Comment (tswast) [inline bigframes/core/compile/compiled.py:793]: Maybe worth a bug upstream in ibis?

File: bigframes/core/compile/compiled.py [hunk 3]
```diff
@@ -815,8 +825,24 @@ def project_window_op(
         window_spec: a specification of the window over which to apply the operator
         output_name: the id to assign to the output of the operator, by default will replace input col if distinct output id not provided
         never_skip_nulls: will disable null skipping for operators that would otherwise do so
-        skip_reproject_unsafe: skips the reprojection step, can be used when performing many non-dependent window operations, user responsible for not nesting window expressions, or using outputs as join, filter or aggregation keys before a reprojection
         """
+        # Cannot nest analytic expressions, so reproject to cte first if needed.
+        # Also ibis cannot window literals, so need to reproject those (even though this is legal in googlesql)
```
Comment (tswast) [inline bigframes/core/compile/compiled.py:830]: Same here. Let's link to an ibis bug.

File: bigframes/core/compile/compiled.py [hunk 4]
```diff
@@ -1034,6 +1057,9 @@ def _to_ibis_expr(
         return table
 
     def filter(self, predicate: ex.Expression) -> OrderedIR:
+        if any(map(is_window, map(self._get_ibis_column, predicate.unbound_variables))):
+            # ibis doesn't support qualify syntax
```
Comment (tswast) [inline bigframes/core/compile/compiled.py:1061]: Might be worth an issue, too. I think there is some support in ibis for `having`, for example (though it's not really possible with the pandas API) if I remember correctly from my work on F1. https://ibis-project.org/tutorials/ibis-for-sql-users#having-clause

Comment (TrevorBergeron) [reply bigframes/core/compile/compiled.py]: created bug and referenced in comment

File: bigframes/core/compile/compiled.py (hunk 3 above)
Comment (TrevorBergeron) [reply bigframes/core/compile/compiled.py:832]: created bug and referenced in comment

File: bigframes/core/compile/compiled.py (hunk 2 above)
Comment (TrevorBergeron) [reply bigframes/core/compile/compiled.py:794]: created bug and referenced in comment

File: bigframes/core/compile/compiled.py (hunk 1 above)
Comment (TrevorBergeron) [reply bigframes/core/compile/compiled.py:349]: This is not a fix, previously we would proactively reproject after windowing. With this new implementation, instead, we are maximally deferring, so need to check later on wherever windowed expressions are not allowed.

**Review by tswast — APPROVED** [review]

**Pull request merged**